### PR TITLE
[SYCL][NFC] Optimize getKernelNamesUsingAssert

### PR DESCRIPTION
**Pull request by maksimsab**
Now it traverses reversed call graph by BFS algorithm from
__devicelib_assert_fail function up to SPIR kernels.

### CODE REVIEW

File: llvm/tools/sycl-post-link/sycl-post-link.cpp [hunk 1]
```diff
@@ -352,124 +355,81 @@ void groupEntryPoints(const Module &M, EntryPointGroupMap &EntryPointsGroups,
...
-      break;
-    case Assert_Indirect:
-      HasIndirectlyCalledAssert = true;
-      break;
-    case No_Assert:
-      break;
+  Optional<const Function *> DevicelibAssertFailFunction;
+  std::vector<StringRef> SPIRKernelNames;
+  // This loop finds all SPIR kernel's names and __devicelib_assert_fail
+  // function if it is present.
```
Comment (bader) [inline llvm/tools/sycl-post-link/sycl-post-link.cpp:409]: I suggest we optimize this function:

```c++
auto DevicelibAssertFailFunction = M.getFunction("__devicelib_assert_fail");
if (!DevicelibAssertFailFunction)
  return {};

auto TraverseResult =
      TraverseCGToFindSPIRKernels(*DevicelibAssertFailFunction);
// Here I would return `true` if no "referenced-indirectly" functions are reached to make code more readable
if (TraverseResult.first)
  return TraverseResult.second;

// Here we reached "referenced-indirectly", so we need to find all kernels and return them.
std::vector<StringRef> SPIRKernelNames;
for (const Function &F : M) {
  if (F.getCallingConv() == CallingConv::SPIR_KERNEL)
    SPIRKernelNames.push_back(F.getName());
}
return SPIRKernelNames;
```

This way to go over all function **only if** there are indirectly callable functions invoking assert function. This should be a rare case.

File: llvm/tools/sycl-post-link/sycl-post-link.cpp [hunk 2]
```diff
@@ -352,124 +355,81 @@ void groupEntryPoints(const Module &M, EntryPointGroupMap &EntryPointsGroups,
...
+// This function traverses over reversed call graph by BFS algorithm.
+// It means that an edge links some function @func with functions
+// which contain call of function @func.It starts from
+// @StartingFunction and lifts up until it reach all reachable functions
+// or it reaches some function containing "referenced-indirectly" attribute.
+// If it reaches "referenced-indirectly" attribute than it returns true and
+// an empty list.
+// Otherwise, it returns false and a list of reached SPIR kernel function's
+// names.
+std::pair<bool, std::vector<StringRef>>
```
Comment (bader) [inline llvm/tools/sycl-post-link/sycl-post-link.cpp:367]: Why not `Optional`?

File: llvm/test/tools/sycl-post-link/assert-indirect-with-split-2.ll [hunk 3]
```diff
@@ -85,6 +78,13 @@ entry:
   ret void
 }
 
+; CHECK-NOT: main_TU0_kernel1
+define dso_local spir_kernel void @main_TU0_kernel1() #0 {
+entry:
+  call spir_func void @_Z4foo1v()
+  ret void
+}
+
```
Comment (AlexeySachkov) [inline llvm/test/tools/sycl-post-link/assert-indirect-with-split-2.ll:87]: Why this change is needed?

File: llvm/test/tools/sycl-post-link/assert-property-2.ll [hunk 4]
```diff
@@ -328,14 +319,23 @@ entry:
...
-; CHECK: _ZTSZZ4mainENK3$_0clERN2cl4sycl7handlerEE7Kernel8
+; CHECK-DAG: _ZTSZZ4mainENK3$_0clERN2cl4sycl7handlerEE7Kernel8
 ; Function Attrs: convergent norecurse mustprogress
 define weak_odr dso_local spir_kernel void @"_ZTSZZ4mainENK3$_0clERN2cl4sycl7handlerEE7Kernel8"() local_unnamed_addr #0 {
   call spir_func void @_Z1Gv()
   call spir_func void @_Z1Hv()
   ret void
 }
 
+; CHECK-NOT: _ZTSZZ4mainENK3$_0clERN2cl4sycl7handlerEE7Kernel6
```
Comment (AlexeySachkov) [inline llvm/test/tools/sycl-post-link/assert-property-2.ll:330]: Please note that `CHECK-NOT` is [bounded](https://llvm.org/docs/CommandGuide/FileCheck.html#the-check-not-directive) by surrounding `CHECK` directives. If you want to ensure that some patters is not present in the file at all, then you should use [`--implicit-check-not`](https://llvm.org/docs/CommandGuide/FileCheck.html#cmdoption-filecheck-implicit-check-not)

File: llvm/tools/sycl-post-link/sycl-post-link.cpp [hunk 5]
```diff
@@ -352,124 +355,81 @@ void groupEntryPoints(const Module &M, EntryPointGroupMap &EntryPointsGroups,
...
-                  isIndirectlyCalledInGraph.end(),
-                  F) != isIndirectlyCalledInGraph.end())
-      HasIndirectlyCalledAttr = true;
-    else if (F->hasFnAttribute("referenced-indirectly")) {
-      HasIndirectlyCalledAttr = true;
-      isIndirectlyCalledInGraph.push_back(F);
-    }
+// This function traverses over reversed call graph by BFS algorithm.
+// It means that an edge links some function @func with functions
+// which contain call of function @func.It starts from
```
Comment (AlexeySachkov) [inline llvm/tools/sycl-post-link/sycl-post-link.cpp:360]:
```suggestion
// which contain call of function @func. It starts from
```

File: llvm/tools/sycl-post-link/sycl-post-link.cpp [hunk 6]
```diff
@@ -352,124 +355,81 @@ void groupEntryPoints(const Module &M, EntryPointGroupMap &EntryPointsGroups,
...
+  std::vector<StringRef> KernelNames;
+
+  while (!FunctionsToVisit.empty()) {
+    const Function *F = FunctionsToVisit.front();
+    FunctionsToVisit.pop();
+
+    // It is possible that we insert some particular function several
+    // times in functionsToVisit queue.
+    if (VisitedFunctions.find(F) != VisitedFunctions.end())
+      continue;
```
Comment (AlexeySachkov) [inline llvm/tools/sycl-post-link/sycl-post-link.cpp:381]:
```suggestion
    auto InsertionResult = VisitedFunctions.insert(F);
    if (!InsertionResult.second)
      continue;
```

File: llvm/tools/sycl-post-link/sycl-post-link.cpp [hunk 7]
```diff
@@ -352,124 +355,81 @@ void groupEntryPoints(const Module &M, EntryPointGroupMap &EntryPointsGroups,
...
-      if (!isa<CallBase>(&I))
-        continue;
+    VisitedFunctions.insert(F);
 
-      const Function *CF = cast<CallBase>(&I)->getCalledFunction();
-      if (!CF)
+    for (const auto *U : F->users()) {
+      const Instruction *I = cast<const Instruction>(U);
+      const Function *ParentF = I->getFunction();
+      if (VisitedFunctions.find(ParentF) != VisitedFunctions.end())
```
Comment (AlexeySachkov) [inline llvm/tools/sycl-post-link/sycl-post-link.cpp:388]:
```suggestion
      if (VisitedFunctions.count(ParentF))
```

File: llvm/test/tools/sycl-post-link/assert-indirect-with-split-2.ll (hunk 3 above)
Comment (maksimsab) [reply llvm/test/tools/sycl-post-link/assert-indirect-with-split-2.ll:87]: New approach lead to a change in the order of kernel names. Because of that I replaced CHECK: with CHECK-DAG:. And it is not allowed to mix CHECK-DAG with CHECK-NOT. That's why I moved some function from DAG sections.

File: llvm/tools/sycl-post-link/sycl-post-link.cpp (hunk 2 above)
Comment (maksimsab) [reply llvm/tools/sycl-post-link/sycl-post-link.cpp]: replaced with Optional.

File: llvm/tools/sycl-post-link/sycl-post-link.cpp [hunk 8]
```diff
@@ -399,39 +398,30 @@ TraverseCGToFindSPIRKernels(const Function *StartingFunction) {
...
-    // If assert is met in some indirectly callable function than
-    // we return all kernels in Module due to the current assert's design.
-    Result = std::move(SPIRKernelNames);
-  } else {
-    Result = std::move(TraverseResult.second);
+      TraverseCGToFindSPIRKernels(DevicelibAssertFailFunction);
+
+  if (TraverseResult.hasValue()) {
+    return std::move(*TraverseResult);
+  }
```
Comment (bader) [inline llvm/tools/sycl-post-link/sycl-post-link.cpp:414]: Please, keep code style consistent and aligned with LLVM coding style guide.
```suggestion
  if (TraverseResult.hasValue())
    return std::move(*TraverseResult);
```

**Review by bader — DISMISSED** [review]
LGTM, just one nit.

**Review by bader — DISMISSED** [review]

File: llvm/tools/sycl-post-link/sycl-post-link.cpp [hunk 9]
```diff
@@ -352,126 +355,71 @@ void groupEntryPoints(const Module &M, EntryPointGroupMap &EntryPointsGroups,
...
+    // It is possible that we insert some particular function several
+    // times in functionsToVisit queue.
+    if (!InsertionResult.second)
+      continue;
 
-      const Function *CF = cast<CallBase>(&I)->getCalledFunction();
-      if (!CF)
+    for (const auto *U : F->users()) {
+      const Instruction *I = cast<const Instruction>(U);
+      const Function *ParentF = I->getFunction();
```
Comment (AlexeySachkov) [inline llvm/tools/sycl-post-link/sycl-post-link.cpp:386]: I'm concerned that there could be other uses of function, which are not instructions, but [`ConstantExpr`](https://llvm.org/doxygen/classllvm_1_1ConstantExpr.html). To generate such LLVM IR we can try to cast one function pointer type into another before doing a call. There is an example of such LLVM IR in [the translator repo](https://github.com/KhronosGroup/SPIRV-LLVM-Translator/blob/master/test/constexpr_vector.ll)

Another concern is that theoretically, an Instruction can be located in module scope,  i.e. outside of a function. Therefore, `I->getFunction()` will return `nullptr`. There is an example of such LLVM IR in [the translator repo](https://github.com/KhronosGroup/SPIRV-LLVM-Translator/blob/master/test/transcoding/SPV_INTEL_function_pointers/const-function-pointer.ll) as well. Note: I don't have good idea about how to generate such LLVM IR out of a SYCL app if you want to replicate it yourself.

**Review by AlexeySachkov — COMMENTED** [review]
Overall LGTM, but I'm concerned that there could be a couple of corner cases not accounted for, related to function pointers

File: llvm/test/tools/sycl-post-link/assert-property-2.ll (hunk 4 above)
Comment (maksimsab) [reply llvm/test/tools/sycl-post-link/assert-property-2.ll]: Unfortunately, today I learnt that `implicit-check-not` option doesn't work with `CHECK-DAG`. At least, in a way which I got from documentation.
For example, consider the following inputs:
file:
```
a
b
c
```
ll:
```
; CHECK-DAG: a
; CHECK-DAG: c
```

If you run `FileCheck ll -input-file=./file --implicit-check-not b` then you get exit code 0 while I would expect to get non-zero.

File: llvm/tools/sycl-post-link/sycl-post-link.cpp (hunk 9 above)
Comment (maksimsab) [reply llvm/tools/sycl-post-link/sycl-post-link.cpp]: Added `dyn_cast` to `CallInst`.

**Comment by maksimsab** [comment]
I did that kernel' names print in a sorted order so that it would be easier to check them with simple `CHECK:`. You might have an approach better than mine.

File: llvm/test/tools/sycl-post-link/assert-property-2.ll (hunk 4 above)
Comment (AlexeySachkov) [reply llvm/test/tools/sycl-post-link/assert-property-2.ll]: > Unfortunately, today I learnt that implicit-check-not option doesn't work with CHECK-DAG. At least, in a way which I got from documentation.

Yeah, I see it to now.

> I did that kernel' names print in a sorted order so that it would be easier to check them with simple CHECK:. You might have an approach better than mine.

We shouldn't be doing extra actions like sorting strings just to make our unit-tests pass. Better solution would be to still work on tests: let's just have two separate FileCheck runs on the same input, but with different prefixes ([`--check-prefix`](https://llvm.org/docs/CommandGuide/FileCheck.html#cmdoption-filecheck-check-prefix)). The first one will check that expected names are present in the result through `CHECK-DAG`, the second one will only check that unexpected names are not present in the result.

File: llvm/tools/sycl-post-link/sycl-post-link.cpp (hunk 9 above)
Comment (AlexeySachkov) [reply llvm/tools/sycl-post-link/sycl-post-link.cpp]: Please note that even though you have fixed a nullptr dereference, functionally we still have an issue, i.e. we can say that there are no assert uses whilst there are through more complex usages of function pointers.

That is probably a minor problem at the moment, because function pointers are not widely used, but we should leave a `FIXME` here, at least, mentioning the use cases we don't analyze properly.

Comment (maksimsab) [reply llvm/tools/sycl-post-link/sycl-post-link.cpp]: Currently, there is a kernel `empty_kernel` in assert-indirect-with-split-2 which checks that traversing doesn't go in such cases.

File: llvm/test/tools/sycl-post-link/assert-property-2.ll (hunk 4 above)
Comment (maksimsab) [reply llvm/test/tools/sycl-post-link/assert-property-2.ll]: Thank you! Done that way.

**Review by AlexeySachkov — APPROVED** [review]

**Comment by AlexeySachkov** [comment]
Two failing tests were fixed about an hour ago in intel/llvm-test-suite#683 and intel/llvm-test-suite#684. Failures were not caused by this patch, but rather by misconfiguration of those tests
